master/custom_steps.py

```diff
@@ -130,6 +131,27 @@ def run(self):
         return status
 
 
+# Like FileUpload, but if the dest file already exists,
+# just log that to stdio and do nothing. Useful when the
+# filename contains (eg) a git commit or SHA that uniquely
+# identifies the file version.
+class FileUploadIfNotExist(FileUpload):
+    name = 'file-upload-if-not-exist'
+
+    def run(self):
+        masterdest = os.path.expanduser(self.masterdest)
+        if os.path.isfile(masterdest) and os.path.getsize(masterdest) > 0:
+            # TODO: this requires @defer.inlineCallbacks, but adding that
+            # will make the non-skip-upload path fail because we have no callbacks.
+            # How to resolve? Not sure. Oh well.
+            #
+            # stdio.addStdout(f"File {repr(masterdest)} already exists on dest, skipping upload!")
+            # yield stdio.finish()
+            return SUCCESS
+
+        return super().run()
```

Comment:
```suggestion
        yield from super().run()
```
When using inline callbacks, use this instead of return when forwarding to another callback-y function. 

Not sure if this resolves the above-mentioned issue.

Reply:
Yeah, I think this should be yield-from.

https://github.com/buildbot/buildbot/blob/master/master%2Fbuildbot%2Fsteps%2Ftransfer.py#L118